A92-xds-ext-authz.md

```diff
@@ -0,0 +1,343 @@
+message](https://github.com/envoyproxy/envoy/blob/cdd19052348f7f6d85910605d957ba4fe0538aec/api/envoy/extensions/filters/network/http_connection_manager/v3/http_connection_manager.proto#L1229)
+in the HCM config and in the [`FilterConfig` wrapper
+message](https://github.com/envoyproxy/envoy/blob/cdd19052348f7f6d85910605d957ba4fe0538aec/api/envoy/config/route/v3/route_components.proto#L2562)
+that can be used in `typed_per_filter_config` fields.
+
+Note that the most specific `typed_per_filter_config` will be used.  For
+example, if there is an override config at the virtual host level that
+disables the filter but then another config at the route level that does
+not disable the filter, the filter will be enabled.
+
+### The ext_authz Protocol
+
+This section describes the ext_authz protocol.
+
+#### Constructing the ext_authz Request
+
+The [`AttributeContext`
+message](https://github.com/envoyproxy/envoy/blob/cdd19052348f7f6d85910605d957ba4fe0538aec/api/envoy/service/auth/v3/attribute_context.proto#L43)
+sent to the server will be populated as follows:
+- [source](https://github.com/envoyproxy/envoy/blob/cdd19052348f7f6d85910605d957ba4fe0538aec/api/envoy/service/auth/v3/attribute_context.proto#L192):
+  Will be set only on gRPC server side.  Inside it:
+  - [address](https://github.com/envoyproxy/envoy/blob/cdd19052348f7f6d85910605d957ba4fe0538aec/api/envoy/service/auth/v3/attribute_context.proto#L58)
+    will be set to the peer address of the connection that the request
+    came in on.
+  - [principal](https://github.com/envoyproxy/envoy/blob/cdd19052348f7f6d85910605d957ba4fe0538aec/api/envoy/service/auth/v3/attribute_context.proto#L82):
+    If TLS is used and the client provided a valid certificate, this will be
+    set to the cert's first URI SAN if set, otherwise the cert's first DNS
+    SAN if set, otherwise the subject field of the certificate in RFC
+    2253 format.  If TLS is not used or the client did not provide a
+    cert, this field will be unset.
+  - [certificate](https://github.com/envoyproxy/envoy/blob/cdd19052348f7f6d85910605d957ba4fe0538aec/api/envoy/service/auth/v3/attribute_context.proto#L86):
+    This will be populated if the `include_peer_certificate` config
+    field is set to true.
+  - service, labels: Will *not* be set.
+- [destination](https://github.com/envoyproxy/envoy/blob/cdd19052348f7f6d85910605d957ba4fe0538aec/api/envoy/service/auth/v3/attribute_context.proto#L197):
+  Will be set only on gRPC server side.  Inside it:
+  - [address](https://github.com/envoyproxy/envoy/blob/cdd19052348f7f6d85910605d957ba4fe0538aec/api/envoy/service/auth/v3/attribute_context.proto#L58)
+    will be set to the local address of the connection that the request
+    came in on.
+  - [principal](https://github.com/envoyproxy/envoy/blob/cdd19052348f7f6d85910605d957ba4fe0538aec/api/envoy/service/auth/v3/attribute_context.proto#L82):
+    If TLS is used, this will be set to the server's cert's first URI SAN
```

Comment:
The `tls` package in Go does not provide a way to retrieve the local certificate used in the TLS handshake.  We had filed an issue for this with the Go team ages ago, but not much progress there: https://github.com/golang/go/issues/24673

If the configured channel credentials contain a certificate provider though, we would be able to retrieve the local certs from the provider. But the provider *could* return more than one cert (if the server is serving multiple domain names and expecting the client to specify an SNI during the TLS handshake). Do other languages have a mechanism to retrieve the actual cert used for the handshake?

Reply:
In C-core, we control the TLS code, so I think we'll be able to handle this.

I'm not sure how hard this will be in Java and Go.  If we can't support it, I'm open to dropping it from the gRFC.

I'd like input from @ejona86, @dfawley, @matthewstevenson88, and @gtcooke94 on this.